ql/src/test/org/apache/hadoop/hive/ql/udf/ptf/TestValueBoundaryScanner.java

```diff
@@ -0,0 +1,187 @@
+    LongWritable w2 = new LongWritable(2);
+
+    Assert.assertTrue(scanner.isEqual(w1, w1));
+
+    Assert.assertFalse(scanner.isEqual(w1, w2));
+    Assert.assertFalse(scanner.isEqual(w2, w1));
+
+    Assert.assertFalse(scanner.isEqual(null, w2));
+    Assert.assertFalse(scanner.isEqual(w1, null));
+
+    Assert.assertTrue(scanner.isEqual(null, null));
+  }
+
+  @Test
+  public void testHiveDecimalEquals() {
+    PTFExpressionDef argDef = new PTFExpressionDef();
+    argDef.setOI(PrimitiveObjectInspectorFactory.writableHiveDecimalObjectInspector);
+
+    HiveDecimalValueBoundaryScanner scanner =
+        new HiveDecimalValueBoundaryScanner(null, null, new OrderExpressionDef(argDef), false);
+    HiveDecimalWritable w1 = new HiveDecimalWritable(1);
+    HiveDecimalWritable w2 = new HiveDecimalWritable(2);
+
+    Assert.assertTrue(scanner.isEqual(w1, w1));
+
+    Assert.assertFalse(scanner.isEqual(w1, w2));
+    Assert.assertFalse(scanner.isEqual(w2, w1));
+
+    Assert.assertFalse(scanner.isEqual(null, w2));
+    Assert.assertFalse(scanner.isEqual(w1, null));
+
+    Assert.assertTrue(scanner.isEqual(null, null));
+  }
+
+  @Test
+  public void testDateEquals() {
+    PTFExpressionDef argDef = new PTFExpressionDef();
+    argDef.setOI(PrimitiveObjectInspectorFactory.writableDateObjectInspector);
+
+    DateValueBoundaryScanner scanner =
+        new DateValueBoundaryScanner(null, null, new OrderExpressionDef(argDef), false);
+    Date date = new Date();
+    date.setTimeInMillis(1000);
+    DateWritableV2 w1 = new DateWritableV2(date);
+    DateWritableV2 w2 = new DateWritableV2(date);
+    DateWritableV2 w3 = new DateWritableV2(); // empty
+
+    Assert.assertTrue(scanner.isEqual(w1, w2));
+    Assert.assertTrue(scanner.isEqual(w2, w1));
+
+    // empty == epoch
+    Assert.assertTrue(scanner.isEqual(w3, new DateWritableV2(new Date())));
+    // empty == another
+    Assert.assertTrue(scanner.isEqual(w1, w3));
+
+    Assert.assertFalse(scanner.isEqual(null, w2));
+    Assert.assertFalse(scanner.isEqual(w1, null));
+
+    Assert.assertTrue(scanner.isEqual(null, null));
+  }
+
+  @Test
+  public void testTimestampEquals() {
+    PTFExpressionDef argDef = new PTFExpressionDef();
+    argDef.setOI(PrimitiveObjectInspectorFactory.writableTimestampObjectInspector);
+
+    TimestampValueBoundaryScanner scanner =
+        new TimestampValueBoundaryScanner(null, null, new OrderExpressionDef(argDef), false);
+    Timestamp ts = new Timestamp();
+    ts.setTimeInMillis(1000);
+
+    TimestampWritableV2 w1 = new TimestampWritableV2(ts);
+    TimestampWritableV2 w2 = new TimestampWritableV2(ts);
+    TimestampWritableV2 w3 = new TimestampWritableV2(); // empty
+
+    Assert.assertTrue(scanner.isEqual(w1, w2));
+    Assert.assertTrue(scanner.isEqual(w2, w1));
+
+    // empty == epoch
+    Assert.assertTrue(scanner.isEqual(w3, new TimestampWritableV2(new Timestamp())));
+    // empty == another
+    Assert.assertFalse(scanner.isEqual(w1, w3));
```

Comment:
This is the opposite behaviour of Date?

Reply:
hm, good catch let me double check

Reply:
this is correct, but comment was incorrect, fixed in:
https://github.com/apache/hive/pull/1950/commits/589126936825ea6652787457dcff6ca038d37f6a